File: backend/app/entities/us_market_theme.py
from __future__ import annotations

from sqlalchemy import CheckConstraint, Float, ForeignKey, Index, Integer, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from backend.app.core.database import Base


class UsThemeGroup(Base):
    __tablename__ = "us_theme_groups"
    __table_args__ = (UniqueConstraint("name", name="uq_us_theme_groups_name"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=100)
    active: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    created_at: Mapped[str] = mapped_column(Text, nullable=False)
    updated_at: Mapped[str] = mapped_column(Text, nullable=False)


class UsTheme(Base):
    __tablename__ = "us_themes"
    __table_args__ = (
        UniqueConstraint("theme_group_id", "name", name="uq_us_themes_group_name"),
        Index("idx_us_themes_group_active", "theme_group_id", "active"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    theme_group_id: Mapped[int] = mapped_column(ForeignKey("us_theme_groups.id", ondelete="RESTRICT"), nullable=False)
    name: Mapped[str] = mapped_column(Text, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    keywords: Mapped[str] = mapped_column(Text, nullable=False, default="")
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=100)
    active: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    created_at: Mapped[str] = mapped_column(Text, nullable=False)
    updated_at: Mapped[str] = mapped_column(Text, nullable=False)


class UsThemeStock(Base):
    __tablename__ = "us_theme_stocks"
    __table_args__ = (
        UniqueConstraint("theme_id", "us_stock_id", name="uq_us_theme_stocks_theme_stock"),
        CheckConstraint("role IN ('LEADER','CORE','RELATED','ETF')", name="ck_us_theme_stocks_role"),
        Index("idx_us_theme_stocks_theme_active", "theme_id", "active"),
        Index("idx_us_theme_stocks_stock_active", "us_stock_id", "active"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    theme_id: Mapped[int] = mapped_column(ForeignKey("us_themes.id", ondelete="CASCADE"), nullable=False)
    us_stock_id: Mapped[int] = mapped_column(ForeignKey("us_stocks.id", ondelete="RESTRICT"), nullable=False)
    role: Mapped[str] = mapped_column(Text, nullable=False, default="RELATED")
    is_representative: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=100)
    active: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    created_at: Mapped[str] = mapped_column(Text, nullable=False)
    updated_at: Mapped[str] = mapped_column(Text, nullable=False)


class UsThemeDailyReturn(Base):
    __tablename__ = "us_theme_daily_returns"
    __table_args__ = (
        UniqueConstraint("theme_id", "trade_date", name="uq_us_theme_daily_returns_theme_date"),
        Index("idx_us_theme_daily_returns_date", "trade_date"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    theme_id: Mapped[int] = mapped_column(ForeignKey("us_themes.id", ondelete="CASCADE"), nullable=False)
    trade_date: Mapped[str] = mapped_column(Text, nullable=False)
    simple_return: Mapped[float] = mapped_column(Float, nullable=False)
    theme_strength: Mapped[float] = mapped_column(Float, nullable=False)
    trimmed_mean_return: Mapped[float] = mapped_column(Float, nullable=False)
    median_return: Mapped[float] = mapped_column(Float, nullable=False)
    breadth_ratio: Mapped[float] = mapped_column(Float, nullable=False)
    valid_stock_count: Mapped[int] = mapped_column(Integer, nullable=False)
    up_count: Mapped[int] = mapped_column(Integer, nullable=False)
    down_count: Mapped[int] = mapped_column(Integer, nullable=False)
    flat_count: Mapped[int] = mapped_column(Integer, nullable=False)
    created_at: Mapped[str] = mapped_column(Text, nullable=False)
    updated_at: Mapped[str] = mapped_column(Text, nullable=False)

File: frontend/src/types/usMarketTheme.ts
export type UsThemeStockRole = "LEADER" | "CORE" | "RELATED" | "ETF";

export type UsThemeSummary = { theme_groups: number; themes: number; active_themes: number; linked_stocks: number };

export type UsThemeGroup = {
  id: number; name: string; description: string | null; sort_order: number; active: number;
  theme_count: number; active_theme_count: number; linked_stock_count: number; created_at: string; updated_at: string;
};

export type UsTheme = {
  id: number; theme_group_id: number; theme_group_name: string; name: string; description: string | null;
  keywords: string[]; sort_order: number; active: number; linked_stock_count: number;
  representative_symbols: string[]; created_at: string; updated_at: string;
  latest_return_date: string | null; latest_simple_return: number | null; latest_theme_strength: number | null; latest_breadth_ratio: number | null;
};

export type UsThemeStock = {
  mapping_id: number; theme_id: number; us_stock_id: number; symbol: string; name: string | null; name_ko: string | null;
  exchange: string; stock_type: string; naver_code: string | null; role: UsThemeStockRole;
  is_representative: number; sort_order: number; active: number; created_at: string; updated_at: string;
};

export type UsThemeGroupInput = { name: string; description?: string | null; sort_order: number; active: number };
export type UsThemeInput = { theme_group_id: number; name: string; description?: string | null; keywords: string[]; sort_order: number; active: number };
export type UsThemeStockInput = { us_stock_id: number; role: UsThemeStockRole; is_representative: number; sort_order: number };
export type UsStockCharts = { stock_id: number; naver_code: string | null; day: string | null; week: string | null; month: string | null; available: boolean };
export type UsThemeReturnItem = {
  theme_id: number; theme_group_name: string; theme_name: string; trade_date: string | null;
  simple_return: number | null; theme_strength: number | null; trimmed_mean_return: number | null; median_return: number | null;
  breadth_ratio: number | null; valid_stock_count: number; up_count: number; down_count: number; flat_count: number;
};
export type UsThemeReturnList = { latest_date: string | null; items: UsThemeReturnItem[] };
export type UsThemeTrendPoint = { trade_date: string; simple_return: number; theme_strength: number; rolling_30d_simple_return: number; rolling_30d_theme_strength: number; rolling_30d_valid_count: number; breadth_ratio: number; valid_stock_count: number; up_count: number };
export type UsThemeTrendItem = { theme_id: number; theme_group_id: number; theme_group_name: string; theme_name: string; active: number; points: UsThemeTrendPoint[] };
export type UsThemeTrend = { period: 20 | 30 | 60; dates: string[]; items: UsThemeTrendItem[] };
export type UsThemeReturnDetail = {
  theme_id: number; theme_name: string; theme_group_name: string; description: string | null; active: number;
  trade_date: string | null; simple_return: number | null; theme_strength: number | null; breadth_ratio: number | null;
  valid_stock_count: number; eligible_stock_count: number; linked_stock_count: number; up_count: number; down_count: number; flat_count: number;
  aggregate: UsThemeReturnItem | null;
  stocks: Array<{
    us_stock_id: number; symbol: string; name: string | null; name_ko: string | null; exchange: string; stock_type: string;
    naver_code: string | null; role: UsThemeStockRole; is_representative: number; sort_order: number; active: number;
    return_rate: number | null; daily_return: number | null; close_price: number | null; previous_close: number | null;
  }>;
};
export type UsMarketRefreshResponse = { price: import("@/types/usStock").UsPriceCollectionResponse; returns: { processed_theme_count: number; processed_date_count: number; upserted_count: number; skipped_count: number; date_from: string | null; date_to: string | null; message: string }; message: string };
